### PR TITLE
Set credentials charset of authentication converter

### LINKED ISSUE #7835: BasicAuthenticationFilter ignores credentials charset
### Summary

`BasicAuthenticationConverter` in `BasicAuthenticationFilter` ignores given credentials charset.

### Actual Behavior

We try to override the credential charset using an `ObjectPostProcessor`:
```java
httpSecurity
     .httpBasic()
     .addObjectPostProcessor(
          new ObjectPostProcessor<BasicAuthenticationFilter>() {
               @Override
                public <O extends BasicAuthenticationFilter> O postProcess(O filter) {
                    filter.setCredentialsCharset("ISO-8859-1"); // <-- try to override default charset
                    return filter;
                }
           }
      );
```
Actual implementation of `BasicAuthenticationFilter#setCredentialsCharset`:

```java
public void setCredentialsCharset(String credentialsCharset) {
    Assert.hasText(credentialsCharset, "credentialsCharset cannot be null or empty");
    this.credentialsCharset = credentialsCharset;
}
```
However, the given `credentialCharset` is never used in `BasicAuthenticationFilter` and especially not in `BasicAuthenticationFilter#BasicAuthenticationConverter`. So, the token is read with the default charset UTF-8 and not as needed with ISO-8859-1.

### Expected Behavior

Corrected `BasicAuthenticationFilter#setCredentialsCharset`:

```java
public void setCredentialsCharset(String credentialsCharset) {
    Assert.hasText(credentialsCharset, "credentialsCharset cannot be null or empty");
    this.credentialsCharset = credentialsCharset; // <-- is never used
    authenticationConverter.setCredentialsCharset(credentialsCharset); // <-- bugfix
}
```

### Version

v5.2.0 - v5.2.1

### PR DESCRIPTION
**Pull request by peterkeller**
Fix #7835 including unit tests

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@peterkeller Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7846)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-security&pullRequestId=7846) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@peterkeller Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7846)!

File: web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java [hunk 1]
```diff
@@ -320,4 +320,105 @@ public void skippedOnErrorDispatch() throws Exception {
 
 		assertThat(response.getStatus()).isEqualTo(200);
 	}
+
+	@Test
+	public void testCredentialsWithUmlautUsingCharset_Utf_8() throws Exception {
```
Comment (eleftherias) [inline web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java]: Although this test name follows the same pattern as other tests in this class, we would like any new tests to follow the convention *methodWhenConditionThenResult*.
A possible name could be:
```suggestion
	public void doFilterWhenTokenAndFilterCharsetMatchThenAuthenticated() throws Exception {
```

File: web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java [hunk 2]
```diff
@@ -320,4 +320,105 @@ public void skippedOnErrorDispatch() throws Exception {
...
+
+		manager = mock(AuthenticationManager.class);
+		when(manager.authenticate(rodRequest)).thenReturn(rod);
+		when(manager.authenticate(not(eq(rodRequest)))).thenThrow(new BadCredentialsException(""));
+
+		filter = new BasicAuthenticationFilter(manager, new BasicAuthenticationEntryPoint());
+
+		String token = "rod:äöü";
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.addHeader("Authorization", "Basic " + new String(Base64.encodeBase64(token.getBytes("UTF-8"))));
```
[the remark this answers is not in the document]
Comment (peterkeller) [reply web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java]: Thanks for your feedback. I adjusted the code.

File: web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java [hunk 3]
```diff
@@ -320,4 +320,105 @@ public void skippedOnErrorDispatch() throws Exception {
...
+		manager = mock(AuthenticationManager.class);
+		when(manager.authenticate(rodRequest)).thenReturn(rod);
+		when(manager.authenticate(not(eq(rodRequest)))).thenThrow(new BadCredentialsException(""));
+
+		filter = new BasicAuthenticationFilter(manager, new BasicAuthenticationEntryPoint());
+		filter.setCredentialsCharset("ISO-8859-1");
+
+		String token = "rod:äöü";
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.addHeader("Authorization", "Basic " + new String(Base64.encodeBase64(token.getBytes("ISO-8859-1"))));
```
[the remark this answers is not in the document]
Comment (peterkeller) [reply web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java]: I adjusted the code.

File: web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java (hunk 1 above)
Comment (peterkeller) [reply web/src/test/java/org/springframework/security/web/authentication/www/BasicAuthenticationFilterTests.java]: I adjusted the code.

**Comment by eleftherias** [comment]
Thanks for the PR @peterkeller! This is now merged into master and backported to 5.2.2.

**Pull request merged**